### PR TITLE
Add support for dbt test macros

### LINKED ISSUE #4185: Support for dbt test macros
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### Description

Currently sqlfluff does not support dbt tests folder. Adding a [custom generic test](https://docs.getdbt.com/guides/legacy/writing-custom-generic-tests) and running lint checks with `sqlfluff lint --config .sqlfluff` 

`{% test test_name %} test_body {% endtest %} `

throws an error

`sqlfluff failure to parse jinja template: encountered unknown tag | 'test'`

Request to support `tests` jinjas

Slack Thread: https://getdbt.slack.com/archives/CV2KM39KR/p1669133058572489

### Use case

_No response_

### Dialect

dbt

### SQLFluff Configuration
```
[sqlfluff]
dialect = bigquery
templater = dbt
processes = 0
exclude_rules = L032, L034

[sqlfluff:indentation]
# See https://docs.sqlfluff.com/en/stable/indentation.html
indented_joins = False
indented_ctes = False
indented_using_on = True
template_blocks_indent = True

[sqlfluff:layout:type:comma]
line_position = trailing

[sqlfluff:templater]
unwrap_wrapped_queries = True

[sqlfluff:templater:dbt]
apply_dbt_builtins = True
# this arg is used only for jinja templater
load_macros_from_path = macros

# Some rules can be configured directly from the config common to other rules
[sqlfluff:rules]
tab_space_size = 4
max_line_length = 120
indent_unit = space
allow_scalar = True
single_table_references = consistent
unquoted_identifiers_policy = all

[sqlfluff:rules:L010]
# Keywords
capitalisation_policy = upper
# Comma separated list of words to ignore for this rule
ignore_words = None

[sqlfluff:rules:L014]
# Unquoted identifiers
extended_capitalisation_policy = lower
# Comma separated list of words to ignore for this rule
ignore_words = None

[sqlfluff:rules:L030]
# Function names
extended_capitalisation_policy = upper
# Comma separated list of words to ignore for this rule
ignore_words = None

[sqlfluff:rules:L040]
# Null & Boolean Literals
capitalisation_policy = upper
# Comma separated list of words to ignore for this rule
ignore_words = None

[sqlfluff:rules:L063]
# Data Types
extended_capitalisation_policy = upper
# Comma separated list of words to ignore for this rule
ignore_words = None
```

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by pdebelak**
### Brief summary of the change made

When using the jinja templater with `apply_dbt_builtins`, this adds support for the dbt test tag. This mostly comes up when also setting `load_macros_from_path = macros` and putting these custom test macros in that macros path as sqlfluff becomes unable to do anything in that case.

fixes #4185 

### Are there any other side effects of this change that we should be aware of?

No

### Pull Request checklist
- [X] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: test/fixtures/templater/jinja_c_dbt/dbt_builtins_test.sql [hunk 1]
```diff
@@ -0,0 +1,11 @@
+{% test my_cool_test(model, column_name, kwarg1=none, kwarg2=none) %}
+SELECT {{ column_name }} FROM {{ model }}
```
Comment (WittierDinosaur) [inline test/fixtures/templater/jinja_c_dbt/dbt_builtins_test.sql:2]: The change broadly looks pretty sensible, but i'm not close to this part of the project. Why does this produce no SQL?

Comment (pdebelak) [reply test/fixtures/templater/jinja_c_dbt/dbt_builtins_test.sql:2]: @WittierDinosaur - these tests are basically like macros. You can see [the dbt docs](https://docs.getdbt.com/guides/best-practices/writing-custom-generic-tests) for details, but you define them like this and then refer to them in your various dbt schema yaml files to run the tests. So defining the test produces no sql itself.

Comment (WittierDinosaur) [reply test/fixtures/templater/jinja_c_dbt/dbt_builtins_test.sql:2]: Ah gotcha, that makes sense

**Review by WittierDinosaur — APPROVED** [review]
LGTM

**Pull request merged**